corehq/ex-submodules/pillowtop/pillow/interface.py

```diff
@@ -100,33 +103,125 @@ def run(self):
 
+    @property
+    def batch_processors(self):
+        if self.processor_chunk_size > 0:
+            return [processor for processor in self.processors if processor.supports_batch_processing]
+        else:
+            return []
+
+    @property
+    def serial_processors(self):
+        if self.processor_chunk_size > 0:
+            return [processor for processor in self.processors if not processor.supports_batch_processing]
+        else:
+            return self.processors
+
     def process_changes(self, since, forever):
-        """
-        Process changes from the changes stream.
-        """
+        # Pass changes from the feed to processors
         context = PillowRuntimeContext(changes_seen=0)
+        min_wait_seconds = 30
         try:
+            changes_chunk = []
+            last_process_time = datetime.utcnow()
             for change in self.get_change_feed().iter_changes(since=since or None, forever=forever):
+                context.changes_seen += 1
                 if change:
-                    context.changes_seen += 1
-                    self.process_with_error_handling(change, context)
+                    # process for all serial_processors
+                    self.process_with_error_handling(change)
+                    # process for all batch_processors
+                    if self.batch_processors:
+                        changes_chunk.append(change)
+                        chunk_full = len(changes_chunk) == self.processor_chunk_size
+                        time_elapsed = (datetime.utcnow() - last_process_time).seconds > min_wait_seconds
+                        if chunk_full or time_elapsed:
+                            last_process_time = datetime.utcnow()
+                            self.batch_process_with_error_handling(changes_chunk, context)
+                            changes_chunk = []
+                    else:
+                        self._update_checkpoint(change, context)
                 else:
                     self._update_checkpoint(None, None)
+            self.batch_process_with_error_handling(changes_chunk, context)
         except PillowtopCheckpointReset:
+            # finish processing any remaining chunk
+            self.batch_process_with_error_handling(changes_chunk, context)
             self.process_changes(since=self.get_last_checkpoint_sequence(), forever=forever)
 
-    def process_with_error_handling(self, change, context):
+    def batch_process_with_error_handling(self, changes_chunk, context):
+        """
+        Passes given changes_chunk to the processor for chunked processing
+            If there is an exception in chunked processing, falls back
+            to passing changes one by one to the processor
+        """
+        for processor in self.batch_processors:
+            if not changes_chunk:
+                return
+            retry_changes = set()
+            timer = TimingContext()
+            with timer:
```

Comment:
This is timing per processor but we're recording the metrics for the pillow as a whole. I do think it's useful to have it be per processor so we should add a tag to the datadog metric so we can separate it out